src/allocator.rs

```diff
@@ -332,13 +336,36 @@ impl Allocator {
 
     pub fn new_pair(&mut self, first: NodePtr, rest: NodePtr) -> Result<NodePtr, EvalErr> {
         let idx = self.pair_vec.len();
-        if idx == MAX_NUM_PAIRS {
+        if idx >= self.max_num_pairs {
```

Comment:
note that we're checking this *before* we add the new pair. i.e. `idx == max_num_pairs` is a valid state. It's just the last valid, we can't add another pair after this.

Reply:
Done